include/circt/Dialect/HW/HWOpInterfaces.td

```diff
@@ -265,6 +265,19 @@ def InnerSymbol : OpInterface<"InnerSymbolOpInterface"> {
               circt::hw::InnerSymbolTable::getInnerSymbolAttrName());
       }]
     >,
+    StaticInterfaceMethod<"Returns whether per-field symbols are supported for this operation type.",
+      "bool", "supportsPerFieldSymbols", (ins), [{}], /*defaultImplementation=*/[{
```

Comment:
At first glance, I was not sure of when would per-field symbol be not supported. Seems like earlier per field symbol did not make any sense, if any op had more than one result. This changes with RefType results. 
If you have a better explanation of why this method can return false, you can add it to the comments, else this is good.

Reply:
I don't have any concrete use cases in mind, but other than being an unnecessary extension point I think having a separate "are per-field symbols supported" query is useful for readability / semantic clarity, if perhaps a bit overkill given our current known needs.  The default implementation is "does the inner symbol target a specific result" for the reason you say :+1: .

I don't expect anything to change this behavior anytime soon, you're not wrong :+1:.

---

> Seems like earlier per field symbol did not make any sense, if any op had more than one result.

Pedantically, the "has more than one result" (or less than one result) was incorrect for instances (and in the other direction, for memories, as we've found in our chat above) and this led to things like #4559 and generally required code to "know" the "meaning" per-op and hopefully get it right.  This was a known limitation of "what do inner symbols and per-field inner symbols /mean/" since they were added, where no answer was identified... so until now has been a bit of "inner symbol names what you want it to" (which has merits of its own, tbh).  The documentation is careful to not say what an inner symbol actually identifies specifically (especially per-field), for this reason.  I guess now we can say "ask the operation what it means!" :rofl: .

> This changes with RefType results.

You're 100% right that adding/extending operations to have multiple results (forceable, reftype's) is what motivated addressing this head-on and making explicit what the result target was, since suddenly the adhoc `getNumResults() != 1` checks throughout are not sufficient and can't easily be made to work without FIRRTL-op/FIRRTL-type-specific knowledge, which shouldn't be used in, e.g., HW.  Should've seen that coming, oops.

Design was driven by "what does HW need to know about each op to accurately verify inner symbols on the op?".  And, on a technical note, motivated by the inability to reasonably change our HW-dialect InnerSym verifier to understand what FIRRTL results to check for which ops (our InstanceOp check has been wrong for some time, using `hw` not `firrtl` for this reason-- don't want to link FIRRTL into HW dialect).

----

As for why the default implementation of `supportsPerFieldSymbols` might not be accurate for some future op:

1) May wish to have a "target result" without supporting per-field symbols within it (this returns true, but false is what the op wants).
2) May wish to not have the (top-level) inner symbol target any result, but per-field inner symbols would have some other meaning.  Maybe selecting which result, or some op-specific behavior.

Mostly want to make whether this is true a clearly different and explicit question to ask an operation, instead of inferring whether this is true from other things (without providing mechanism for agreeing/disagreeing with the default).